### PR TITLE
Allow configuring server to provide relative URLS for resources

### LINKED ISSUE #5692: Make it possible to use relative URLs
I'm trying to get bokeh running in a docker/docker-swarm behind a nginx with path to port mapping.
Unfortunately bokeh produces full URL's when creating links which doesn't work in my case. 

Starting bokeh with:
`bokeh serve untitled.py --host=* --disable-index-redirect`

And accessing it over: 
` https://my.bokeh.server:8443/9335bf292e454cf8ba1181befd4d4a99/p5006/`

Produce links like the following:
` <link rel="stylesheet" href="http://127.0.0.1:5006/static/css/bokeh.min.css?v=7246afcfffc127faef7c138bce4742e9" type="text/css" />`

I've tried setting the explicit `--host` and  also tried the `--prefix` option, both doesn't fit my requirements. What would work is a possibility to switch using relative URL's. Means  instead of `http://127.0.0.1:5006/static/`  use `./static/` or even better `static/`. This is the same approach that shiny Apps are using.
For the websocket it could be calculated on the client side
```
      var protocol = 'ws:';
      if (window.location.protocol === 'https:')
        protocol = 'wss:';
       // TODO check for '/' in pathname and pathname is not already decoded
      var defaultPath = window.location.pathname  + '/websocket/';
      var ws = new WebSocket(protocol + '//' + window.location.host + defaultPath);
```

So my enhancement is:
**Please, make it possible to enable relative URL generation.** For example with `--use-relative-urls` 

See also https://groups.google.com/a/continuum.io/forum/#!topic/bokeh/lIrCvXZE18M

### PR DESCRIPTION
**Pull request by bryevdv**
- [x] issues: fixes #5692
- [x] tests added / passed
- [x] release document entry (if new feature or API change)

This PR makes it possible for the server to supply relative URLs for resources, which may be needed for certain kinds of deployments behind proxies. It's not yet clear if the WS code also needs changing, awaiting more input from @dhuebner

### CODE REVIEW

**Comment by havocp** [comment]
I'm not clear on why this is a special case; if it works, we could do it always and get rid of the --host option and request.host checking. The rule would be then that it's a (security AND functional) bug to use request.host, so might want to somehow enforce that by linting for its use or making it throw an exception.

Or if this doesn't work, then we need --host and deployments that don't know their hostname(s) just can't work.

**Comment by bryevdv** [comment]
OK, I am going to expand this PR to try and remove the need for `--hosts` to be set

**Comment by bryevdv** [comment]
I'm going to clean up the commits in this PR, but I want to see all the tests run on this checkpoint.

**Comment by bryevdv** [comment]
ping @mattpap @havocp there may some additional cleanup/docs but I'd appreciate comments/review. 

Is the simple test for no `request.host` usage reasonable? I don't think it's perfect but it should afford more confidence than no check at all. 

This seems to work great. I tried with "native" apps and `push_session` style apps. I also tested some notebook examples including notebook comms, as well as standard examples. 

I need to test "deploy" scenarios (e.g. behind nginx or embedded w/ autoload_server) but it would be much easier to merge this after some review and test those cases with a dev build.

File: bokeh/client/session.py [hunk 1]
```diff
@@ -7,11 +7,36 @@
...
+    elif url.startswith("wss:"):
+        reprotocoled = "https" + url[3:]
+    else:
+        raise ValueError("URL has non-websocket protocol " + url)
+    if not reprotocoled.endswith("/ws"):
+        raise ValueError("websocket URL does not end in /ws")
+    return reprotocoled[:-2]
+
+DEFAULT_SERVER_WEBSOCKET_URL = websocket_url_for_server_url(DEFAULT_SERVER_HTTP_URL)
+
```
Comment (bryevdv) [inline bokeh/client/session.py]: Python client connections do need to construct full websocket URLs, AFAIK? But in that case move that code here under `bokeh.client`.

File: bokeh/command/subcommands/serve.py [hunk 2]
```diff
@@ -495,6 +458,9 @@ def invoke(self, args):
         log_level = getattr(logging, args.log_level.upper())
         logging.basicConfig(level=log_level, format=args.log_format)
 
+        if args.host is not None and len(args.host) > 0:
+            warnings.warn("The --hosts parameter is no longer needed or necessary and is ignored. It will be removed and trigger an error in a future release")
+
```
Comment (bryevdv) [inline bokeh/command/subcommands/serve.py]: `warnings.warn` good? Seems better than a log message.

File: bokehjs/src/coffee/embed.coffee [hunk 3]
```diff
@@ -158,7 +158,13 @@ fill_render_item_from_script_tag = (script, item) ->
...
   logger.info("Will inject Bokeh script tag with params #{JSON.stringify(item)}")
 
-export embed_items = (docs_json, render_items, websocket_url=null) ->
+export embed_items = (docs_json, render_items) ->
+  protocol = 'ws:'
+  if (window.location.protocol == 'https:')
+    protocol = 'wss:'
+  ws_pathname = window.location.pathname.replace(/\/+$/, '') + '/ws'
+  websocket_url = protocol + '//' + window.location.host + ws_pathname
+
```
Comment (bryevdv) [inline bokehjs/src/coffee/embed.coffee]: Here is the code to compute the WS URL in the browser

File: scripts/tests/test_api_report.py [hunk 4]
```diff
@@ -5,7 +5,7 @@
 from ..api_report import diff_versions
 
 
-@pytest.mark.api_test
+@pytest.mark.skip(reason="version checkouts currently too disruptive")
 def test_working_pipeline():
     first = diff_versions("0.11.0", "0.12.0").split("\n")
     with open("scripts/tests/samples/sample_diff.txt") as f:
```
Comment (bryevdv) [inline scripts/tests/test_api_report.py]: Unrelated. Turning off this test as accidentally running is very disruptive to a local dev env

**Comment by bryevdv** [comment]
@dhuebner If I get out a dev build with this PR merged, can you help test it?

**Comment by havocp** [comment]
Looks ok to me on quick scan. The main thing I think the request.host test could miss would be a use of request.host by a library, but I can't think of a library we use that would do that (tornado itself? doesn't generate any urls afaik).  The textual search seems like a good precaution against simple mistakes.

**Comment by dhuebner** [comment]
@bryevdv Sure, with the greatest of pleasure

**Comment by bryevdv** [comment]
@dhuebner brilliant, that's a huge help. I will have a new dev build out sometime this afternoon.

**Comment by bryevdv** [comment]
@dhuebner `dev13` is available via conda:
```
conda install -c bokeh/c/dev bokeh
```

**Comment by marami52** [comment]
Hi @bryevdv I'm a colleague of @dhuebner but not a developer though! I just tested your package but get 404 error. In the console, I have 400 (Bad request) error for favicon.ico

In fact it's looking at https://my.bokeh.server:8443/favicon.ico instead of: 

https://my.bokeh.server:8443/9335bf292e454cf8ba1181befd4d4a99/p5006/ which is my current link. Am I missing something or it needs more changes? 

Thank you for this great work!

**Comment by bryevdv** [comment]
@marami52 it's a little hard to say without knowing more about your deployment. Bokeh doesn't serve any templates that request a favicon. Are you embedding using `autoload_server` The original issue mentioned path to port mapping, it definitely seems possible there needs to be a different configuration to accommodate the relative URLs

**Comment by bryevdv** [comment]
@havocp I think the use of relative URLs means that `autoload_server` can't work **except** behind some kind of proxy. Consider this. I just ran:
```
bokeh serve --prefix=5010 --show --prefix=foo sliders.py
```
Then I run `python -m http.server` to server this file:
```
<!DOCTYPE html>
<html>
<head>
    <title>Crap</title>
</head>
<body>
<script
    src="http://localhost:5006/foo/sliders/autoload.js?bokeh-autoload-element=9d4a6493-c611-406c-aea3-b14f519dad0b"
    id="9d4a6493-c611-406c-aea3-b14f519dad0b"
    data-bokeh-model-id=""
    data-bokeh-doc-id=""
></script>
</body>
```
Now, the when the page is loaded and that script is executed, now it uses relative URLs, which fail because the simple built-in python server doesn't know anything about forwarding:
```
[Log] Bokeh: injecting script tag for BokehJS library:  – "static/js/bokeh.min.js?v=ad8c62292415d93f8d4eaa900bcd3d6c" (autoload.js, line 51)
[Log] Bokeh: injecting script tag for BokehJS library:  – "static/js/bokeh-widgets.min.js?v=3affff8390d2171fd375a3226ab52045" (autoload.js, line 51)
[Error] Failed to load resource: the server responded with a status of 404 (File not found) (bokeh-widgets.min.js, line 0)
[Error] Failed to load resource: the server responded with a status of 404 (File not found) (bokeh.min.js, line 0)
```
I will try this with an nginx reverse proxy config. I *think* that will work. But is there anything to do about the above situation with trying to embed in "simple" http server?

**Comment by marami52** [comment]
@bryevdv thanks for explanation. After this stage is beyond my knowledge! :) I'll leave it with @dhuebner to get back to you with more details of our setup.

**Comment by bryevdv** [comment]
OK I have found some changes that seem to get everything working behind a reverse proxy. But there is still the issue described above when trying to embed using `autoload_server` in a simple non-proxied case. I suppose in that case an option would be to let the resources URL be specified by the creator explicitly. That would still allow Bokeh to avoid every using the HTML request host header. I think that's reasonable (@havocp your input welcome) but I want to try and make a new single PR for that rather than build on this. Accordingly I am reverting this PR for now.

**Comment by dhuebner** [comment]
@bryevdv Sorry, I missed the github notification.
I've just tried the new dev version it looks better now.  The problem I currently have is the ws connection:
`WebSocket connection to 'ws://192.168.99.100:8080/p5006/myapp/ws?bokeh-protocol-version=1.0&bokeh-session-id=cI4ULkWJwFdLPF9djSffG7S2HUj6moEQZVgwwO9eMwGB' failed: Error during WebSocket handshake: Unexpected response code: 400`

I need to check the nginx settings I think. I will report my findings later.

**Comment by dhuebner** [comment]
@bryevdv  Now it works! Had to to tweak our nginx for the new bokeh ws. And after adding

`--allow-websocket-origin=*` all works as expected. Great! Thanks a lot!

**Comment by bryevdv** [comment]
@dhuebner OK that's good to know, as I mentioned I reverted this for the time being, because it was not working for me:

* behind a proxy with a `--prefix` also used
* without a proxy just embedded e.g. in local flask app

But I intend to pick it up and have an updated PR soon: I have some stashed code to deal with the first issue, and I think the second can be addressed by offering a new command line option to explicitly provide the host and port to connect to for resources. I think this covers all the cases, and also means there is no use of `request.host` which I regard as a big simplification for everyone. It would make the one case, an embed in web app without a proxy, require one more configuration (the resources URL).

Is there agreement that that is a reasonable tradeoff? @bokeh/dev

**Comment by marami52** [comment]
@bryevdv as @dhuebner mentioned before dev13 release was working well but dev14 is not working as you said you reverted that. So, do you have a plan for an updated PR for this issue? the issue is already closed in Bokeh.

**Comment by bryevdv** [comment]
@marami52 @dhuebner I do plan to submit an updated PR, hopefully next week. My intention is for this to be closed for `0.12.5`. I have re-opened the issue in the mean time (should have happened before, it just slipped my mind) For the immediate term if you need this I can only suggest sticking with `dev13`

**Pull request merged**